### PR TITLE
more balanced 16x32, 32x16 and 32x32 decisions

### PR DESCRIPTION
**Pull request by jyrkialakuijala**
Reduces ringing significantly in d16 and worse.
Some reduction in d8.
Finest qualities get changes only in smooth areas that already look
great.

0.13 % improvement on BPP*pnorm
0.58 % on butteraugli distance
0.016 % size reduction

Before:

```
Compr                 Input    Compr            Compr       Compr  Decomp  Butteraugli
Method               Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm   PSNR   QABPP     DCT2     DCT4   DCT4X8      AFV     DCT8  DCT8X16  DCT8X32    DCT16 DCT16X32    DCT32           BPP*pnorm   Errors
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
jxl:d0.8:epf0      17972865  3952043    1.75911542205   1   2.348  13.709   1.57893729   0.52330958503  41.73   2.121 0.000000 0.000911 0.203567 0.020553 0.049902 0.230698 0.000000 0.408836 0.017035 0.068597  0.9205619615324337        0
jxl:d1:epf1        17972865  3450570    1.53590203899   1   2.565  20.029   1.94662952   0.60612205767  40.65   2.128 0.000000 0.000911 0.192140 0.020753 0.049001 0.201234 0.000000 0.396715 0.028686 0.110872  0.9309441042592400        0
jxl:d2:epf3        17972865  2246256    0.99984326372   1   2.850  16.446   3.41899276   0.95125633921  37.50   2.256 0.000000 0.001139 0.159970 0.021308 0.044515 0.151339 0.000000 0.301154 0.098338 0.223284  0.9511072428334978        0
jxl:d3:epf0        17972865  1680140    0.74785628223   1   2.758  24.550   4.57588768   1.27704026411  35.28   2.324 0.000000 0.001139 0.134923 0.020033 0.041385 0.134367 0.000000 0.233724 0.175026 0.261970  0.9550425841796468        0
jxl:d4:epf3        17972865  1390103    0.61875633072   1   2.409  14.737   5.85040092   1.56284952646  34.11   2.425 0.000000 0.001139 0.111054 0.018039 0.034800 0.120786 0.000000 0.200152 0.228554 0.289147  0.9670230384675962        0
jxl:d8:epf1        17972865   813211    0.36197278508   1   2.823  24.289   9.37444782   2.51584719998  30.89   2.406 0.000000 0.001367 0.058484 0.011665 0.018783 0.088738 0.000000 0.132865 0.326921 0.367145  0.9106682178227494        0
jxl:d16:epf3       17972865   475322    0.21157316877   1   2.872  14.535  21.55422401   4.42386021454  28.62   2.641 0.000000 0.001367 0.016729 0.003927 0.003998 0.049290 0.000000 0.072942 0.332077 0.527130  0.9359701238049873        0
jxl:d24:epf3       17972865   358029    0.15936424159   1   2.590  14.324  21.47615433   5.77418376695  27.34   2.649 0.000000 0.001367 0.007673 0.001791 0.001741 0.033907 0.000000 0.051647 0.293705 0.616068  0.9201984168450851        0
jxl:d32:epf3       17972865   292046    0.12999418846   1   2.847  13.934  25.84580421   7.02234695199  26.51   2.668 0.000000 0.016636 0.003781 0.000897 0.000957 0.026137 0.000000 0.036677 0.250632 0.672359  0.9128642931176300        0
Aggregate:         17972865  1125217    0.50085125627 ---   2.667  16.944   6.81123365   1.86408749564  33.21   2.393 0.000000 0.001552 0.051750 0.008541 0.013921 0.091328 0.000000 0.151461 0.135466 0.279207  0.9336305639835538        0
```

After:

```
Compr                 Input    Compr            Compr       Compr  Decomp  Butteraugli
Method               Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm   PSNR   QABPP     DCT2     DCT4   DCT4X8      AFV     DCT8  DCT8X16  DCT8X32    DCT16 DCT16X32    DCT32           BPP*pnorm   Errors
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
jxl:d0.8:epf0      17972865  3951942    1.75907046539   1   2.324  13.807   1.57893729   0.52331089591  41.73   2.121 0.000000 0.000911 0.203567 0.020553 0.049902 0.230626 0.000000 0.408822 0.017234 0.068483  0.9205407412221729        0
jxl:d1:epf1        17972865  3450706    1.53596257469   1   2.538  20.125   1.94662952   0.60610525355  40.65   2.128 0.000000 0.000911 0.192140 0.020753 0.049001 0.201177 0.000000 0.396829 0.029199 0.110303  0.9309549857768020        0
jxl:d2:epf3        17972865  2246506    0.99995454258   1   2.826  16.524   3.41899276   0.95121670445  37.50   2.246 0.000000 0.001139 0.159978 0.021315 0.044529 0.151353 0.000000 0.301396 0.104035 0.217301  0.9511734645939139        0
jxl:d3:epf0        17972865  1679780    0.74769604067   1   2.780  24.514   4.76437998   1.27694160442  35.28   2.321 0.000000 0.001139 0.135008 0.020033 0.041445 0.134325 0.000000 0.233283 0.181521 0.255816  0.9547641817919317        0
jxl:d4:epf3        17972865  1390051    0.61873318472   1   2.398  14.544   5.85040092   1.56297457360  34.10   2.430 0.000000 0.001139 0.110922 0.018061 0.034779 0.120629 0.000000 0.198984 0.239920 0.279233  0.9670642355616933        0
jxl:d8:epf1        17972865   811812    0.36135006856   1   2.788  23.422   9.37444782   2.51424949430  30.89   2.407 0.000000 0.001367 0.058295 0.011626 0.018769 0.088538 0.000000 0.128891 0.354582 0.343899  0.9085242271468007        0
jxl:d16:epf3       17972865   475248    0.21154023023   1   2.836  13.781  19.64723206   4.41670512959  28.63   2.675 0.000000 0.001367 0.016718 0.003920 0.004023 0.048898 0.000000 0.067344 0.381702 0.483488  0.9343108199727461        0
jxl:d24:epf3       17972865   358126    0.15940741779   1   2.588  13.989  21.47615433   5.75762783145  27.36   2.626 0.000000 0.001367 0.007627 0.001805 0.001705 0.033558 0.000000 0.045195 0.354981 0.561657  0.9178085852275816        0
jxl:d32:epf3       17972865   292118    0.13002623677   1   2.796  14.169  25.83151817   6.98935655673  26.53   2.639 0.000000 0.016636 0.003810 0.000904 0.000961 0.026058 0.000000 0.031706 0.315754 0.612251  0.9087997305448716        0
Aggregate:         17972865  1125031    0.50076852951 ---   2.646  16.767   6.77137520   1.86203845214  33.21   2.391 0.000000 0.001552 0.051733 0.008552 0.013905 0.091067 0.000000 0.144941 0.148250 0.266164  0.9324502575761613        0
```

### CODE REVIEW

File: lib/jxl/enc_ac_strategy.cc [hunk 1]
```diff
@@ -272,6 +272,60 @@ namespace HWY_NAMESPACE {
...
+  }
+  if (y % 8 == 0) {
+    // Nothing crosses 64x64 boundaries, and the memory on the other side
+    // of the 64x64 block may still uninitialized.
+    return false;
+  }
+  end_x = std::min(end_x, ac_strategy.xsize());
+  AcStrategyRow row = ac_strategy.ConstRow(y);
+  for (size_t x = start_x; x < end_x;) {
+    if (row[x].IsMultiblock()) {
```
Comment (veluca93) [inline lib/jxl/enc_ac_strategy.cc:289]: Why is this line here at all?

File: lib/jxl/enc_ac_strategy.cc [hunk 2]
```diff
@@ -659,6 +713,149 @@ void FindBest16X16(size_t bx, size_t by, size_t cx, size_t cy,
...
+  const AcStrategy acs32X32 = AcStrategy::FromRawStrategy(acs_raw32X32);
+  AcStrategyRow row0 = ac_strategy->ConstRow(by + cy + 0);
+  AcStrategyRow row1 = ac_strategy->ConstRow(by + cy + 2);
+  {
+    bool has32X32 = row0[bx + cx + 0].RawStrategy() == acs_raw32X32;
+    if (has32X32) {
+      return;
+    }
+  }
+  if (MultiBlockTransformCrossesHorizontalBoundary(*ac_strategy, bx + cx,
```
Comment (veluca93) [inline lib/jxl/enc_ac_strategy.cc:746]: When can this happen?

File: lib/jxl/enc_ac_strategy.cc [hunk 3]
```diff
@@ -659,6 +713,149 @@ void FindBest16X16(size_t bx, size_t by, size_t cx, size_t cy,
...
+  const bool allow_16X32 = !MultiBlockTransformCrossesHorizontalBoundary(
+      *ac_strategy, bx + cx, by + cy + 2, bx + cx + 4);
+  // Current entropies aggregated on 16x16 resolution.
+  float entropy[2][2] = {};
+  for (size_t dy = 0; dy < 4; ++dy) {
+    for (size_t dx = 0; dx < 4; ++dx) {
+      entropy[dy >> 1][dx >> 1] += entropy_estimate[(cy + dy) * 8 + (cx + dx)];
+    }
+  }
+  float try32X16_0 = std::numeric_limits<float>::max();
```
Comment (veluca93) [inline lib/jxl/enc_ac_strategy.cc:767]: This is a bad name for the variable

File: lib/jxl/enc_ac_strategy.cc [hunk 4]
```diff
@@ -659,6 +713,149 @@ void FindBest16X16(size_t bx, size_t by, size_t cx, size_t cy,
...
+      EstimateEntropy(acs32X32, (bx + cx + 0) * 8, (by + cy + 0) * 8, config,
+                      cmap_factors, block, scratch_space, quantized);
+
+  // Test if this block should have 32X16 or 16X32 transforms,
+  // because it can have only one or the other.
+  float cost32x16 = std::min(try32X16_0, entropy[0][0] + entropy[1][0]) +
+                    std::min(try32X16_1, entropy[0][1] + entropy[1][1]);
+  float cost16x32 = std::min(try16X32_0, entropy[0][0] + entropy[0][1]) +
+                    std::min(try16X32_1, entropy[1][0] + entropy[1][1]);
+  if (try32X32 < cost32x16 && try32X32 < cost16x32) {
```
Comment (veluca93) [inline lib/jxl/enc_ac_strategy.cc:810]: It sounds to me that this whole if structure should be refactored to not have basically the same code repeated 4 times (and I assume that the 16x16 function is very similar too)

File: lib/jxl/enc_ac_strategy.cc (hunk 1 above)
Comment (jyrkialakuijala) [reply lib/jxl/enc_ac_strategy.cc:289]: I wasn't aware that first block concept is well-defined for single blocks. Thank you!

File: lib/jxl/enc_ac_strategy.cc (hunk 2 above)
Comment (jyrkialakuijala) [reply lib/jxl/enc_ac_strategy.cc:746]: It is important when this code will be applied to floating blocks later, similar to the FindBest16X16 is used today.

As the code is used today, this will not trigger.

File: lib/jxl/enc_ac_strategy.cc (hunk 3 above)
Comment (jyrkialakuijala) [reply lib/jxl/enc_ac_strategy.cc:767]: renamed more descriptive (with left/right/top/bottom instead of 0/1)

File: lib/jxl/enc_ac_strategy.cc (hunk 4 above)
Comment (jyrkialakuijala) [reply lib/jxl/enc_ac_strategy.cc:810]: I'll generalize this in a next step to 32X64,64X32 and 64X64, and then see how to simplify it.

I'll try to keep refactorings and new development in separate PRs so that there is more clarity of no adverse effects from refactorings.

Comment (jyrkialakuijala) [reply lib/jxl/enc_ac_strategy.cc:810]: Adding a TODO about this

File: lib/jxl/enc_ac_strategy.cc (hunk 2 above)
Comment (veluca93) [reply lib/jxl/enc_ac_strategy.cc:746]: Can we add a comment for this then?

File: lib/jxl/enc_ac_strategy.cc (hunk 3 above)
Comment (veluca93) [reply lib/jxl/enc_ac_strategy.cc:767]: I was more complaining about "try"

File: lib/jxl/enc_ac_strategy.cc (hunk 4 above)
Comment (veluca93) [reply lib/jxl/enc_ac_strategy.cc:810]: Fair enough, as long as the final state doesn't have this repetition.

Note that the argument of "no adverse effects from refactorings" doesn't apply to not repeating the code 4 times with slightly different constants in the lines below, though - it would be better to make those parts a lambda.

Comment (jyrkialakuijala) [reply lib/jxl/enc_ac_strategy.cc:810]: changed my mind and did a small refactoring already, but think that more is needed -- however that I'd like to postpone to the point when this code is first generalized to all sizes

File: lib/jxl/enc_ac_strategy.cc (hunk 3 above)
Comment (jyrkialakuijala) [reply lib/jxl/enc_ac_strategy.cc]: both fixed

File: lib/jxl/enc_ac_strategy.cc (hunk 2 above)
Comment (jyrkialakuijala) [reply lib/jxl/enc_ac_strategy.cc:752]: added comments

**Review by veluca93 — APPROVED** [review]
